File: sprint3/By languages/Java/lesson11code1.java
public static int[] mergeSort(int[] array) {
    if (array.length == 1) {  // базовый случай рекурсии
        return array;
    }

    // запускаем сортировку рекурсивно на левой половине
    int[] left = mergeSort(Arrays.copyOfRange(array, 0, array.length/2));

    // запускаем сортировку рекурсивно на правой половине
    int[] right = mergeSort(Arrays.copyOfRange(array, array.length/2, array.length));

    // заводим массив для результата сортировки
    int[] result = new int[array.length];

    // сливаем результаты
    int l = 0, r = 0, k = 0;
    while (l < left.length && r < right.length) { 
        // выбираем, из какого массива забрать минимальный элемент
        if (left[l] <= right[r]) { 
            result[k] = left[l];
            l++;
        } else {
            result[k] = right[r];
            r++;
        }
        k++;
    }

    // Если один массив закончился раньше, чем второй, то
    // переносим оставшиеся элементы второго массива в результирующий
    while (l < left.length) { 
        result[k] = left[l];   // перенеси оставшиеся элементы left в result
        l++;
        k++;
    }
    while (r < right.length) { 
        result[k] = right[r];  // перенеси оставшиеся элементы right в result
        r++;
        k++;
    }

    return result;
}

File: sprint3/By languages/swift/lesson11code1.swift
func mergeSort(_ array: [Int]) -> [Int] {
    guard array.count > 1 else {
        return array
    }
    
    let middle = array.count / 2
    // запускаем сортировку рекурсивно на левой половине
    let leftArray = mergeSort(Array(array[..<middle]))
    // запускаем сортировку рекурсивно на правой половине
    let rightArray = mergeSort(Array(array[middle...]))
    
    return merge(leftArray, rightArray)
}

func merge(_ left: [Int], _ right: [Int]) -> [Int] {
    var mergedArray: [Int] = []
    var l = 0
    var r = 0
    
    while l < left.count && r < right.count {
        if left[l] < right[r] {
            mergedArray.append(left[l])
            l += 1
        } else {
            mergedArray.append(right[r])
            r += 1
        }
    }
    
    mergedArray.append(contentsOf: left[l...])
    mergedArray.append(contentsOf: right[r...])
    
    return mergedArray
}

let unsortedArray = [11, 2, 9, 7, 1]
let sortedArray = mergeSort(unsortedArray)
print(sortedArray)